### PR TITLE
rustdoc: render more cross-crate HRTBs properly

### PR DESCRIPTION
**Pull request by fmease**
Follow-up to #102439.
Render the `for<>` parameter lists of cross-crate higher-rank trait bounds (in where-clauses and in `impl Trait`).

I've added a new field `bound_params` to `clean::WherePredicate::EqPredicate` (mirroring its sibling variant `BoundPredicate`). However, I had to box the existing fields since `EqPredicate` used to be the largest variant (128 bytes on 64-bit systems) and it would only have gotten bigger).
Not sure if you like that approach. As an alternative, I could pass the uncleaned `ty::Predicate` alongside the cleaned `WherePredicate` to the various re-sugaring methods (similar to what `clean::AutoTraitFinder::param_env_to_generics` does).

I haven't yet added the HTML & JSON rendering code for the newly added `bound_params` field since I am waiting for your opinion. Those two rendering code paths should actually be unreachable in practice given the fact(?) that we re-sugar all equality predicates to associated type bindings (and arbitrary equality predicates are not part of the Rust surface language at the time of this writing).

If you agree with storing `bound_params` in `EqPredicate`, I think I can use it to greatly simplify the `clean::auto_trait` module (by also using `simplify::merge_bounds`). Maybe I can do that in any case though.

@rustbot label T-rustdoc A-cross-crate-reexports
r? @GuillaumeGomez

### CODE REVIEW

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Thanks for the PR! The code changes look good to me, but two things before merging:

 1. We need a perf check because of the changes in the `impl_trait` variable in `clean`.
 2. I'd like to have a second opinion about this to be sure I didn't miss anything (so cc @notriddle).

@bors try @rust-timer queue

**Comment by GuillaumeGomez** [comment]
Perf check is down maybe?

**Comment by fmease** [comment]
https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra/topic/bors.20stuck.3F/near/302463516

**Comment by fmease** [comment]
spurious failure (network timeouts, 504)

**Comment by GuillaumeGomez** [comment]
Ah seems like the bot is back. Let's retry and if there is another spurious failure, we'll just wait a bit before trying again.

@bors try @rust-timer queue

File: src/librustdoc/clean/simplify.rs
Comment (fmease) [inline src/librustdoc/clean/simplify.rs:105]:
```suggestion
        trait_ref.generic_params.dedup_by_key(|p| p.name);
```

Note to self: Regardless of whether rust-timer is going to report a performance regression or not, I am going to change this line for increased performance. Contrary to the `sort_unstable_by` line which requires comparison between strings, we definitely don't need to access the underlying value of the interned strings here (the whole point of interned strings, duh! 😅).

Comment (GuillaumeGomez) [reply src/librustdoc/clean/simplify.rs:105]: If they're symbol, you need to access the underlying string as it is otherwise not the sort you might expect. Unless you don't care it's alpha sorted?

**Comment by GuillaumeGomez** [comment]
It seems that your changes have a small impact on perf. The memory impact is bit high though but it's in acceptable boundaries. Please ping me once you updated the `dedup` check so we can re-run a perf check.

**Comment by fmease** [comment]
I suspect some of the calls to `.collect()` / `.extend(_)` or the repeated call to `.clone()` in a loop may also be a culprit. About to investigate for a short while.

**Comment by fmease** [comment]
Okay, I've adjusted the `dedup` code & fixed a small logic bug which may also improve perf. Not sure if you need to wait for CI to pass before requesting a perf run. Anyways,
@rustbot ready

**Review by notriddle — APPROVED** [review]

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Thanks!

@bors r+

**Comment by fmease** [comment]
Hmm, since the same commit results in two contradicting perf results, I'd say it might just be noise.

**Comment by GuillaumeGomez** [comment]
I think so too.

**Pull request merged**
